Refer-A-Friend System prepared for s1980 adopted from archiv s1699

Hikarii opened the pull request:
This is a (nearly) 1:1 copy of the Refer-A-Friend System of the archiv repository at https://github.com/mangos-one/server-old

Taken from
- https://github.com/mangos-one/server-old/commit/00cb8bb4fa9c90e19f612a28f18ad41887f97ee4
- https://github.com/mangos-one/server-old/commit/6a37d17f757075739ec901e222e96fb65bbd6adb

Signed-off-by: Salja salja2012@hotmail.de

Salja committed:
[s1980] Implement Refer-A-Friend System. Thanks all Authors
Taken from
- https://github.com/mangos-one/server-old/commit/00cb8bb4fa9c90e19f612a28f18ad41887f97ee4
- https://github.com/mangos-one/server-old/commit/6a37d17f757075739ec901e222e96fb65bbd6adb

Signed-off-by: Salja <salja2012@hotmail.de>

Hikarii commented on src/game/World.cpp:
```diff
@@ -570,6 +570,12 @@ void World::LoadConfigSettings(bool reload)
+    setConfigMinMax(CONFIG_UINT32_RAF_MAXREFERALS, "RAF.MaxReferals", 5, 0, 15);
+    setConfigMinMax(CONFIG_UINT32_RAF_MAXREFERERS, "RAF.MaxReferers", 5, 0, 15);
+    setConfig(CONFIG_FLOAT_RATE_RAF_XP, "Rate.RAF.XP", 3.0f);
+    setConfig(CONFIG_FLOAT_RATE_RAF_LEVELPERLEVEL, "Rate.RAF.XP", 0.5f);
```
Typo: Should be setConfig(CONFIG_FLOAT_RATE_RAF_LEVELPERLEVEL, "Rate.RAF.LevelPerLevel", 0.5f);

Salja commented:
some more need fixed the patch make a traget bug i think the patch need a full rewirte

Hikarii commented:
You talk about the target bug by summoning a friend by using the corresponding context menu option, while targeting another player, NPC or creature, which leads to fail the spell, right?

Salja commented:
yes thats i mean =) and i think some we can make better in the code some more clean etc

Hikarii commented:
I've some suggestions for the new system. I've read blizzards article about refer a friend and I've found that there are more conditions which are missing in the current system.
- CONFIG_UINT32_RAF_MAXGRANTLEVEL should also affect CONFIG_FLOAT_RATE_RAF_XP. Above the given level there will be no longer bonus xp.
- CONFIG_UINT32_RAF_MAXGRANTLEVEL should also affect the summonig spell. players with level equal or lower than the value can be summoned. Above not.
- If the level difference between the players ar bigger than 3 (former it was 4, which includes tbc), then only the lower one gain bonus xp the other gain normal xp.
- You can only grant a level to a character which is below your own level.
- The bonus xp of the refer a friend system should be shown in the combat log (like the rested bonus).

These informations are from blizzards system for Mist of Pandaria. Whether this matchs the situation back in The Burning Crusade... I don't know.

Commit:
Fix Error with Travis
- Revert my poor implementation of the friend list function to its initial state of the archive.
  Travis error was caused because #include<iterater> wasn't included anywhere. But I think there is a better solution for this function as using std::ostream_iterator<T> like I did.

- Fix typo while loading RAF configuration.

billy1arm commented:
@Hikarii - Do you think you can get this fixed up so that we can attempt testing on it ?

Hikarii commented:
@billy1arm I don't know. I'll try it. What do you want to be fixed? The suggestions I made above?

billy1arm commented:
@Hikarii yes please

Hikarii commented:
@billy1arm Currently the RAF system is in its original state, how it was back in rel 12.1. Hence untested for Rel 19!
I haven't pushed the fix for reputation gain yet. I can do that yet, if you want.
Unfortunately I haven't made any progress in the context menu targeting issue mentioned [above](https://github.com/mangosone/server/pull/13#issuecomment-16573835).

billy1arm commented:
@Hikarii  - could you push the 'reputation gain fix' ?

Hikarii committed:
Fix reputation gain in RAF system
The bonus on reputation, granted by killing creatures and quest rewards, is 10%, not 300%.
